File: kinematics_np/sparke_leg_IK.py
import numpy as np
import leg_transformations as legtf
import base_transformations as basetf

class sparkeLeg():
    def __init__(self, leg_id):
        self.init_variables(leg_id)
        self.initialize_leg_transforms()

    def init_variables(self, leg_id):
        self.x_dir, self.y_dir = legtf.get_dirs(leg_id)
        self.len1 = 0.055
        self.len2 = 0.125
        self.len3 = 0.135

    def initialize_leg_transforms(self):
        self.t_01 = legtf.create_T01(0)
        self.t_12 = legtf.create_T12(0)
        self.t_23 = legtf.create_T23(0)

    def update_Tb0(self, Tm):
        self.t_b0 = legtf.create_Tb0(Tm, self.x_dir, self.y_dir)

    def solve_angles(self, Tm, x_ee, y_ee, z_ee):
        self.update_Tb0(Tm)
        self.solve_theta1(y_ee, z_ee)
        self.solve_theta3(x_ee, z_ee)
        self.solve_theta2(x_ee, z_ee)

    def solve_theta1(self, y_ee, z_ee):
        y0, z0 = self.t_b0[1, 3], self.t_b0[2, 3]
        c = np.sqrt(((z_ee-z0)**2) + ((y_ee-y0)**2))
        b = np.sqrt((c**2)-(self.len1**2))
        thetaA = np.arctan2((z_ee-z0), (y_ee-y0)*self.y_dir)
        thetaB = np.arctan2(-b, self.len1)
        theta1 = thetaB - thetaA
        self.theta1 = round(theta1, 4)

    def solve_theta3(self, x_ee, z_ee):
        t_b1 = self.get_tb1()
        x1, z1 = t_b1[0,3], t_b1[2,3]
        x1_ee = x_ee - x1
        z1_ee = z_ee - z1 #CHECK for which should be first
        a = (x1_ee**2) + (z1_ee**2) - (self.len2**2) - (self.len3**2)
        b = -2*self.len2*self.len3
        theta3 = (np.arccos(a/b)/2)
        self.theta3 = round(theta3, 4)

    def solve_theta2(self, x_ee, z_ee):
        t_b1 = self.get_tb1()
        x1, z1 = t_b1[0,3], t_b1[2,3]
        a = self.len3 * np.sin(self.theta3) * self.x_dir
        b = self.len3 * np.cos(self.theta3)
        x2 = abs(x_ee - a)
        z2 = abs(z_ee + b)
        x1_2 = x1 - x2
        z1_2 = z1 - z2
        theta2 = np.arctan2(z1_2, x1_2)
        self.theta2 = round(theta2, 4)

    def get_tb1(self):
        self.t_01 = legtf.create_T01(self.theta1)
        t_b1 = np.matmul(self.t_b0, self.t_01)
        return t_b1

# def main():
#     Tm = basetf.create_base_transformation(0, 0, 0, 0, 0, 0)
#     x_ee = 0.1080710678
#     y_ee = 0.11
#     z_ee = -0.1838477631
#     leg_test = sparkeLeg(1)
#     leg_test.solve_angles(Tm, x_ee, y_ee, z_ee)
#     print(f'theta1: {leg_test.theta1}')
#     print(f'theta2: {leg_test.theta2}')
#     print(f'theta3: {leg_test.theta3}')


# if __name__ == '__main__':
#     main()

File: src/sparke_leg_IK.cpp
#include <Eigen/Dense>
#include <iostream>
#include <cmath>
#include "leg_transformations.hpp"
#include "base_transformations.hpp"
#include "sparke_leg_IK.hpp"

SparkeLeg::SparkeLeg(int leg_id)
{
    get_dirs(leg_id);
    t_01 = create_T01(0);
    t_12 = create_T12(0);
    t_23 = create_T23(0);
}

SparkeLeg::~SparkeLeg()
{

}

void SparkeLeg::update_Tb0(Eigen::Matrix<float, 4, 4> Tm)
{
    t_b0 = create_Tb0(Tm, x_dir, y_dir);
}

void SparkeLeg::solve_angles(Eigen::Matrix<float, 4, 4> Tm, float x_ee,
                                                     float y_ee, float z_ee)
{
    update_Tb0(Tm);
    solve_theta1(y_ee, z_ee);
    solve_theta3(x_ee, z_ee);
    solve_theta2(x_ee, y_ee);
}

void SparkeLeg::solve_theta1(float y_ee, float z_ee)
{
    float y0, z0, c, b, thetaA, thetaB;
    y_ee = abs(y_ee);
    z_ee = abs(z_ee);
    y0 = abs(t_b0(1,3));
    z0 = abs(t_b0(2,3));
    c = sqrt((pow((z_ee-z0), 2)) + (pow((y_ee-y0), 2)));
    b = sqrt((pow(c,2))-(pow(LEN_1,2)));
    thetaA = atan2(abs(z_ee - z0), abs(y_ee-y0));
    thetaB = atan2(b, LEN_1);
    theta1 = thetaB - thetaA;
}

void SparkeLeg::solve_theta3(float x_ee, float z_ee)
{
    float x1, z1, x1_ee, z1_ee, a, b;
    Eigen::Matrix<float,4,4> t_b1 = get_tb1();
    x1 = abs(t_b1(0,3));
    z1 = abs(t_b1(2,3));
    x1_ee = x_ee - x1;
    z1_ee = z_ee - z1;
    a = pow(x1_ee, 2) + pow(z1_ee, 2) - pow(LEN_2, 2) - pow(LEN_3, 2);
    b = -2 * LEN_2 * LEN_3;
    theta3 = acos(a/b);
}

void SparkeLeg::solve_theta2(float x_ee, float z_ee)
{
    float x1, z1, x1_3, z1_3, alpha, beta, a, b, c;
    Eigen::Matrix<float,4,4> t_b1 = get_tb1();
    x1 = abs(t_b1(0,3));
    z1 = abs(t_b1(2,3));
    x1_3 = abs(x_ee - x1) * -1;
    z1_3 = z_ee - z1;
    alpha = atan2(z1_3, x1_3);
    c = sqrt(pow(x1_3, 2) + pow(z1_3, 2));
    a = pow(LEN_3, 2) - pow(LEN_2, 2) - pow(c, 2);
    b = -2*LEN_2*c;
    beta = acos(a/b);
    theta2 = alpha - beta;
}

Eigen::Matrix<float, 4, 4> SparkeLeg::get_tb1()
{
    t_01 = create_T01(theta1);
    Eigen::Matrix<float,4,4> t_b1 = t_b0 * t_01;
    return t_b1;
}

void SparkeLeg::get_dirs(int leg_id)
{
    if (leg_id==1)
    {
        x_dir = 1;
        y_dir = 1;
    } 
    else if (leg_id == 2)
    {
        x_dir = 1;
        y_dir = -1;
    }
    else if (leg_id == 3)
    {
        x_dir = -1;
        y_dir = 1;
    }
    else if (leg_id == 4)
    {
        x_dir = -1;
        y_dir = -1;
    }
}
